Translation of the page "Lists and Keys" (#112)
reactjs/it.reactjs.org: content/docs/lists-and-keys.md

---
id: lists-and-keys
title: Liste e Chiavi
permalink: docs/lists-and-keys.html
prev: conditional-rendering.html
next: forms.html
---

Prima di iniziare, rivediamo come trasformare le liste in JavaScript.

> Nel codice qui sotto, usiamo la funzione [`map()`](https://developer.mozilla.org/it/docs/Web/JavaScript/Reference/Global_Objects/Array/map) per prendere un array di `numeri` e raddoppiare i loro valori. Assegniamo il nuovo array restituito da `map()` alla variabile `lista` e lo stampiamo a console:

Reviewer: ```suggestion
Nel codice qui sotto, usiamo la funzione [`map()`](https://developer.mozilla.org/it/docs/Web/JavaScript/Reference/Global_Objects/Array/map) per prendere un array di `numeri` e raddoppiarne i valori. Assegniamo il nuovo array restituito da `map()` alla variabile `lista` e lo stampiamo a console:
```
the `array` is the subject here

const lista = numeri.map((numero) => numero * 2);
console.log(lista);
```

Questo codice mostra `[2, 4, 6, 8, 10]` nella console.

Trasformare array in liste di [elementi](/docs/rendering-elements.html) con React è quasi identico.

### Renderizzare Liste di Componenti {#rendering-multiple-components}

Puoi creare liste di elementi e [usarle in JSX](/docs/introducing-jsx.html#embedding-expressions-in-jsx) usando le parentesi graffe `{}`.

Di seguito, eseguiamo un ciclo sull'array `numeri` usando la funzione JavaScript [`map()`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/map). Ritorniamo un elemento `<li>` per ogni elemento dell'array. Infine, assegniamo l'array risultante a `lista`:

```javascript{2-4}
const numeri = [1, 2, 3, 4, 5];
const lista = numeri.map((numero) =>
  <li>{numero}</li>
);
```

Includiamo l'intero array `lista` all'interno di un elemento `<ul>` e [lo renderizziamo nel DOM](/docs/rendering-elements.html#rendering-an-element-into-the-dom):

```javascript{2}
ReactDOM.render(
  <ul>{lista}</ul>,
  document.getElementById('root')
);
```

**[Prova su CodeSandbox](codesandbox://lists-and-keys/1.js)**

> Questo codice visualizza un elenco puntato di numeri tra 1 e 5.

Reviewer: ```suggestion
Questo codice visualizza un elenco puntato di numeri da 1 a 5.
```
Maybe it's me but `tra 1 and 5` sounds like 2,3,4 to me. 🤔

Reviewer 2: Yes, "da 1 a 5" is clear while "tra 1 e 5" leaves you wondering whether 1 and 5 are included or not

Author: Yeah, agree. Good spot!

```js{2}
const listaArticoli = articoli.map((articolo) =>
  <li key={articolo.id}>
    {articolo.testo}
  </li>
);
```

Quando non disponi di identificativi stabili per gli elementi da renderizzare, puoi assegnare l'indice dell'elemento corrente alla chiave come ultima scelta:

```js{2,3}
const listaArticoli = articoli.map((articolo, indice) =>
  // Fallo solo se gli elementi non hanno identificativi stabili
  <li key={indice}>
    {articolo.testo}
  </li>
);
```

Non consigliamo di utilizzare gli indici per le chiavi se l'ordine degli elementi potrebbe cambiare. Potrebbe avere un impatto negativo sulle prestazioni e causare problemi con lo stato dei componenti. Leggi l'articolo di Robin Pokorny per una [spiegazione approfondita sugli impatti negativi dell'uso di un indice come chiave](https://medium.com/@robinpokorny/index-as-a-key-is-an-anti-pattern-e0349aece318). React utilizza gli indici come chiavi se non assegni esplicitamente una chiave per renderizzare gli elementi.

> Qui trovi una [spiegazione approfondita sul perché le chiavi sono necessarie](/docs/reconciliation.html#recursing-on-children) se sei interessato a saperne di più.

Reviewer: ```suggestion
Qui trovi una [spiegazione approfondita sul perché le chiavi sono necessarie](/docs/reconciliation.html#recursing-on-children) se vuoi saperne di più.
```

    // Sbagliato! La chiave deve essere stata specificata qui:
    <Numero valore={numero} />
  );
  return (
    <ul>
      {lista}
    </ul>
  );
}

const numeri = [1, 2, 3, 4, 5];
ReactDOM.render(
  <ListaNumeri numeri={numeri} />,
  document.getElementById('root')
);
```

**Esempio: Corretto Utilizzo della Chiave**

```javascript{2,3,9,10}
function Numero(props) {
  // Corretto! Non è necessario specificare la chiave qui:
  return <li>{props.valore}</li>;
}

function ListaNumeri(props) {
  const numeri = props.numeri;
  const lista = numeri.map((numero) =>
    // Corretto! La chiave deve essere specificata all'interno dell'array.
    <Numero key={numero.toString()}
            valore={numero} />
  );
  return (
    <ul>
      {lista}
    </ul>
  );
}

const numeri = [1, 2, 3, 4, 5];
ReactDOM.render(
  <ListaNumeri numeri={numeri} />,
  document.getElementById('root')
);
```

**[Prova su CodeSandbox](codesandbox://lists-and-keys/3.js)**

È buona regola ricordarsi che che gli elementi all'interno della chiamata `map()` hanno bisogno di chiavi.

### Le Chiavi Devono Essere Uniche Tra Gli Elementi Adiacenti {#keys-must-only-be-unique-among-siblings}

> Chiavi usate all'interno degli array dovrebbero essere uniche tra gli elementi adiacenti. Tuttavia, non hanno bisogno di essere unici a livello globale. Possiamo usare le stesse chiavi quando creiamo due array diversi:

Reviewer: ```suggestion
Chiavi usate all'interno degli array dovrebbero essere uniche tra gli elementi adiacenti. Tuttavia, non hanno bisogno di essere uniche a livello globale. Possiamo usare le stesse chiavi quando creiamo due array diversi:
```

Reviewer 2: ```suggestion
Le chiavi usate all'interno degli array dovrebbero essere uniche tra gli elementi adiacenti. Tuttavia, non hanno bisogno di essere uniche a livello globale. Possiamo usare le stesse chiavi quando creiamo due array diversi:
```

      {contenuto}
    </div>
  );
}

const articoli = [
  {id: 1, titolo: 'Ciao Mondo', testo: 'Benvenuto in imparando React!'},
  {id: 2, titolo: 'Installazione', testo: 'Puoi installare React usando npm.'}
];
ReactDOM.render(
  <Blog articoli={articoli} />,
  document.getElementById('root')
);
```

**[Prova su CodeSandbox](codesandbox://lists-and-keys/4.js)**

Le chiavi servono a React come suggerimento, ma non vengono passate ai componenti. Se hai bisogno di quel valore nel tuo componente, passalo come prop esplicitamente con un nome diverso:

```js{3,4}
const contenuto = articoli.map((articolo) =>
  <Articolo
    key={articolo.id}
    id={articolo.id}
    titolo={articolo.titolo} />
);
```

In questo esempio, il componente `Articolo` può leggere` props.id`, ma non `props.key`.

### Incorporare map() in JSX {#embedding-map-in-jsx}

Nell'esempio di prima abbiamo dichiarato una variabile separata `lista` e l'abbiamo usata nel codice JSX:

```js{3-6}
function ListaNumeri(props) {
  const numeri = props.numeri;
  const lista = numeri.map((numero) =>
    <Numero key={numero.toString()}
            valore={numero} />
  );
  return (
    <ul>
      {lista}
    </ul>
  );
}
```

> JSX consente [incorporare qualsiasi espressione](/docs/introducing-jsx.html#embedding-expressions-in-jsx) in parentesi graffe in modo da poter scrivere direttamente il risultato di `map()`:

Reviewer: ```suggestion
JSX consente di [incorporare qualsiasi espressione](/docs/introducing-jsx.html#embedding-expressions-in-jsx) in parentesi graffe in modo da poter scrivere direttamente il risultato di `map()`:
```

```js{5-8}
function ListaNumeri(props) {
  const numeri = props.numeri;
  return (
    <ul>
      {numeri.map((numero) =>
        <Numero key={numero.toString()}
                valore={numero} />
      )}
    </ul>
  );
}
```

**[Prova su CodeSandbox](codesandbox://lists-and-keys/5.js)**

> A volte questo codice risulta più chiaro, ma questa alternativa può anche essere abusata. Come in JavaScript, spetta a te decidere se vale la pena estrarre una variabile per ragioni di leggibilità del codice. Tieni presente che se vi sono troppi elementi _nested_ nel corpo `map()`, potrebbe essere arrivato il momento di [estrarre un componente](/docs/components-and-props.html#extracting-components).

Reviewer: ```suggestion
A volte questo codice risulta più chiaro, ma questa alternativa può anche essere abusata. Come in JavaScript, spetta a te decidere se vale la pena estrarre una variabile per ragioni di leggibilità del codice. Tieni presente che se vi sono troppi elementi _nested_ (ovvero: annidati) nel corpo `map()`, potrebbe essere arrivato il momento di [estrarre un componente](/docs/components-and-props.html#extracting-components).
```
or something like that
